src/passes/Inlining.cpp

```diff
@@ -262,14 +260,41 @@ struct FunctionInfoScanner
       }
 
       if (shrinks) {
-        info.trivialCall = TrivialCall::Shrinks;
+        info.trivialInstruction = TrivialInstruction::Shrinks;
         return;
       }
 
       if (info.size == call->operands.size() + 1) {
-        // This function body is a call with some trivial (size 1) operands like
-        // LocalGet or Const, so it is a trivial call.
-        info.trivialCall = TrivialCall::MayNotShrink;
+        // This function body is an instruction with some trivial (size 1)
+        // operands like LocalGet or Const, so it is a trivial instruction.
+        info.trivialInstruction = TrivialInstruction::MayNotShrink;
+      }
+
+    } else if (auto* binary = curr->body->dynCast<Binary>()) {
+      info.trivialInstruction = TrivialInstruction::MayNotShrink;
+      if (auto* left = binary->left->dynCast<LocalGet>()) {
+        if (auto* right = binary->right->dynCast<LocalGet>()) {
+          if (right->index > left->index) {
+            info.trivialInstruction = TrivialInstruction::Shrinks;
+            return;
+          }
+        }
+      }
+
+      if (info.size == 3) {
+        // Same as above: if arguments have size 1 we consider it a trivial
+        // instruction.
+        info.trivialInstruction = TrivialInstruction::MayNotShrink;
+      }
+
+    } else if (auto* unary = curr->body->dynCast<Unary>()) {
+      info.trivialInstruction = TrivialInstruction::MayNotShrink;
+      if (unary->value->dynCast<LocalGet>()) {
+        info.trivialInstruction = TrivialInstruction::Shrinks;
+      } else if (info.size == 2) {
+        // Same as above: if the argument has size 1 we consider it a trivial
+        // instruction.
+        info.trivialInstruction = TrivialInstruction::MayNotShrink;
```

Comment:
I believe we can unify the Call, Binary, and Unary paths. Rather than iterate over children manually, there is ChildIterator, allowing you to do
```cpp
for (auto* child : ChildIterator(curr)) {
```

Reply:
Yeah, I think that makes sense. I updated the code to use `ChildIterator`, which simplified it quite a bit, and it's now easy to handle more instructions.

We can support things like `i31.ref`, string instructions, and many others, but we have to manually list them here now in the conditional with `body->is<Call>() || body->is<...>() || ...`. Is there an easier way?

Should I keep adding more instructions in this PR, or in a separate PR?

Reply:
Why not allow all instructions? I can't think of any that I would want to disallow here, but maybe I'm missing something?

From a general point of view, if the function does one operation and just uses local.gets in order to do so, that seems good to inline, no matter what the operation is?

(Actually a Call is the only one that seems risky, if the call is recursive, as we'd end up inlining it to itself many times. But I think we might already have logic to handle that.)

Reply:
Could block instructions with block type operands increase binary sizes when duplicated at the call sites?

Reply:
Actually, maybe I was wrong, as other control flow operations add risk here. If and Try, for example, are larger than a call, so inlining this:
```wat
(if (result i32)
  (local.get $x)
  (then
    (local.get $y)
  )
  (else
    (local.get $z)
  )
)
```
could increase code size. Also, the arms are not both executed, so it's not clear that the optimizer can remove the local operations.

Perhaps we can exit if `Properties::isControlFlowStructure()`?